cassiopeia/tools/branch_length_estimator.py

```diff
@@ -0,0 +1,325 @@
+
+        l2_penalty = 0
+        for (parent, child) in tree.edges():
+            for child_of_child in tree.children(child):
+                edge_length_above = (
+                    r_X_t_variables[parent] - r_X_t_variables[child]
+                )
+                edge_length_below = (
+                    r_X_t_variables[child] - r_X_t_variables[child_of_child]
+                )
+                l2_penalty += (edge_length_above - edge_length_below) ** 2
+        l2_penalty *= l2_regularization
+
+        # # # # # Solve the problem # # # # #
+
+        obj = cp.Maximize(log_likelihood - l2_penalty)
+        prob = cp.Problem(obj, all_constraints)
+
+        f_star = prob.solve(solver="ECOS", verbose=verbose)
+
+        # # # # # Populate the tree with the estimated branch lengths # # # # #
+
+        for node in tree.nodes():
+            tree.set_age(node, age=r_X_t_variables[node].value)
+
+        for (parent, child) in tree.edges():
+            new_edge_length = (
+                r_X_t_variables[parent].value - r_X_t_variables[child].value
+            )
+            tree.set_edge_length(parent, child, length=new_edge_length)
+
+        self.log_likelihood = log_likelihood.value
+        self.log_loss = f_star
+
+    @classmethod
+    def log_likelihood(self, tree: Tree) -> float:
+        r"""
+        The log-likelihood of the given tree under the model.
+        """
+        log_likelihood = 0.0
+        for (parent, child) in tree.edges():
+            edge_length = tree.get_age(parent) - tree.get_age(child)
+            # TODO: hardcoded '0' here...
+            zeros_parent = tree.get_state(parent).count("0")
+            zeros_child = tree.get_state(child).count("0")
+            new_cuts_child = zeros_parent - zeros_child
+            assert new_cuts_child >= 0
+            # Add log-lik for characters that didn't get cut
+            log_likelihood += zeros_child * (-edge_length)
+            # Add log-lik for characters that got cut
+            if edge_length < 1e-8 and new_cuts_child > 0:
+                return -np.inf
+            log_likelihood += new_cuts_child * np.log(1 - np.exp(-edge_length))
+        return log_likelihood
+
+
+class IIDExponentialBLEGridSearchCV(BranchLengthEstimator):
+    r"""
+    Like IIDExponentialBLE but with automatic tuning of hyperparameters.
+
+    This class fits the hyperparameters of IIDExponentialBLE based on
+    character-level held-out log-likelihood. It leaves out one character at a
+    time, fitting the data on all the remaining characters. Thus, the number
+    of models trained by this class is #characters * grid size.
+
+    Args:
+        minimum_branch_lengths: The grid of minimum_branch_length to use.
+        l2_regularizations: The grid of l2_regularization to use.
+        verbose: Verbosity level.
+    """
+
+    def __init__(
+        self,
+        minimum_branch_lengths: Tuple[float] = (0,),
+        l2_regularizations: Tuple[float] = (0,),
```

Comment:
Re: tuples, isn't this the natural parameterization of a grid? E.g. `minimum_branch_lengths  = (0, 1), l2_regularizations  = (0, 10, 100, 1000)`. The grid is given by all possible combinations of both parameters.

Re: hyperparam tuning library: Yes! I agree! I was peeking a bit into sklearn to see if we could reuse anything. I just wanted to get CV up and running for experimentation purposes, so I went the fast route of just coding it myself. We should look into a generic and reusable implementation.

Reply:
Sounds good! I had a total brainfart and was thinking you were using tuples that were of length 2. No reason why you can't use Tuples here rather than lists. 

But yes, let's investigate hyperparameter tuning libraries as I imagine they'll do a lot of good stuff about exploring large parameter regimes intelligently. For example, Google has `Vizier` (https://cloud.google.com/ai-platform/optimizer/docs/overview) that not only explores large parameter spaces but actually optimizes some function at the same time so that the new experiments you do from your grid of parameters are likely improvements.